### ISSUE TITLE
connect: no buffer space available

### ISSUE THREAD

**Issue by zcalusic** [restic/restic#642]
## Output of `restic version`

restic 0.3.0 (v0.3.0-8-g974f2f7)
compiled at 2016-10-13 21:26:38 with go1.7.1 on linux/amd64
## Expected behavior

Backup withour errors.
## Actual behavior

error loading tree: obj.Stat: Head http://minio.server:9000/backup/restic/data/eeebe2cc06b17deaa6802d39ed84f431d3bc34b76e8d0873a312dadb7dd5f539: dial tcp 11.22.33.44:9000: connect: no buffer space available
error loading tree: Get http://minio.server:9000/backup/restic/data/9a0724d69550464a7fed33617fe3bde792e11f4e87f0f68a963240ed86e203fe: dial tcp 11.22.33.44:9000: connect: no buffer space available
error loading tree: Get http://minio.server:9000/backup/restic/data/9a0724d69550464a7fed33617fe3bde792e11f4e87f0f68a963240ed86e203fe: dial tcp 11.22.33.44:9000: connect: no buffer space available
## Steps to reproduce the behavior

Run backup, nothing special.

I'm seeing this errors occasionaly. I don't yet know their effect on the backup consistency. I will run check soon. Backup goes to minio server that is very close and with enough bandwidth. These errors seem intermittent and maybe restic should retry connect after a while if it can? Maybe errors are coming due to too many simultaneous connections? Finally, maybe minio is doing something wrong here (again)?

**Comment by fd0** [comment]
Huh, thanks for reporting this. I haven't seen this error yet, and googling around finds https://groups.google.com/forum/#!topic/golang-nuts/6bUrwr5dto8 which points to connections not being closed.

How much data did you store?

**Comment by zcalusic** [comment]
The political answer would be, much less than I intend to. ;)

This is now 2 servers backing up to the 3rd server with minio S3 repo. Also backup is going simultaneously, from both backup clients at about same time. There's 33GB of data in minio repo, 7400 files in restic data folder.

The plan is to order a storage server with several TB of disk space, exclusively for restic. And then point few dozens of servers there. That's why this ENOBUFS is a bit frightening, it could happen much more often with more servers doing backup.

Maybe not a right place to ask, but what do you think would be the best (most efficient) protocol to backup few dozens servers to the central backup restic repository. S3, or FTP, or NFS, or...?

**Comment by fd0** [comment]
I'd say s3 is probably overkill, you can use sftp (which is very well tested) or use the REST backend (which is not yet fully documented but works, and there's a sample server implementation in `cmds/restic-server`.

**Comment by zcalusic** [comment]
I completely forgot about sample server implementation. Thanks for reminding me. I'll check that next on this (still) test setup, and see how it works. And sftp too, a bit later.

**Comment by zcalusic** [comment]
Before I proceed with testing other backends, here's restic check run on the S3 repo.
```
Create exclusive lock for repository
Load indexes
Check all packs
Check snapshots, trees and blobs
error for tree 4433d5cc:
  tree 4433d5cc: file "system.journal" has nil blob list
error for tree df572db9:
  tree df572db9: file "system.journal" has nil blob list
error for tree de4e7244:
  tree de4e7244: file "system.journal" has nil blob list
error for tree 33887ead:
  tree 33887ead: file "database.sqlite" has nil blob list
error for tree 119cd162:
  tree 119cd162: file "port-2.rrd" has nil blob list
  tree 119cd162: file "port-3.rrd" has nil blob list
  tree 119cd162: file "processor-hr-768.rrd" has nil blob list
Fatal: repository contains errors
restic check  1711.50s user 425.35s system 274% cpu 12:57.56 total
```
I'd say the errors resemble #604 bug, not ENOBUFS.

**Comment by fd0** [comment]
That is #604, the buffer issue is something different.

**Comment by fd0** [comment]
Are you still seeing this error? Any idea what might have caused it?

I'm inclined to close this issue as not reproducible, is that okay for you?

**Comment by zcalusic** [comment]
Believe it or not, this is a very reproducable issue, especially now when I changed backend I'm using (REST now, vs S3 before). It is just manifesting a bit differently. As REST backend is quite a lot faster than S3, this error can now be reliably reproduced in just a few seconds. Some other explanation could be that S3 backend was so slow only because of exactly this issue.

My plan is to attack this issue and fix it properly, as soon as I find some quality time to do it. It is currently not obvious if this is restic or even Go runtime bug.

Stay tuned...

**Comment by fd0** [comment]
Great, thanks for tackling this yourself!

**Comment by zcalusic** [comment]
I can't be 100% sure, but I believe the fix similar to what I proposed in #669 for REST backend would fix this issue in S3 backend. It feels about the same problem, even if a bit differently reported. The trouble is, I deconstructed my minio S3 setup, and don't plan to use it anymore, seeing how much faster the REST backend is (thanks for the advice to test it, BTW :+1:). So, I'm afraid somebody else will have to look into this issue, somebody who plans to use S3 backend.

**Comment by fd0** [comment]
I'm going to apply the same fix as #669 and close this issue, let's see if it turns up again.

**Comment by fd0** [comment]
Ah, I had a look at the code and it looks good (at least to me). Hm. It's a shame you don't have your setup any more ;)

**Comment by yhafri** [comment]
Is this issue fixed by applying the fix as in [#669](https://github.com/restic/restic/pull/669) ?
Or not yet?

**Comment by fd0** [comment]
The code in #669 does not touch the s3 backend code at all. I'm closing this issue as not reproducible, until someone shows up who can help debugging it.

@yhafri did you run into the same issue?
